broker/lib/controllers/FabrikBaseController.js

```diff
@@ -30,6 +33,67 @@ class FabrikBaseController extends BaseController {
     return _.get(config, 'broker_name', 'service-fabrik-broker');
   }
 
+  handleWithResourceLocking(func, operationType, lastOperationCall) {
+    return (req, res, next) => {
+      return Promise.try(() => {
+          if (req.manager.name === CONST.INSTANCE_TYPE.DIRECTOR && config.enable_service_fabrik_v2) {
+            // Acquire lock for this instance
+            return lockManager.lock(req.params.instance_id, {
+              lockedResourceDetails: {
+                resourceGroup: CONST.APISERVER.RESOURCE_GROUPS.DEPLOYMENT,
+                resourceType: CONST.APISERVER.RESOURCE_TYPES.DIRECTOR,
+                resourceId: req.params.instance_id,
+                operation: operationType ? operationType : req.query.operation.type // This is for the last operation call
+              }
+            });
+          }
+        })
+        .then(() => this._handleWithUnlock(func, operationType, lastOperationCall, req, res, next))
+        .catch(err => {
+          if (lastOperationCall && err instanceof DeploymentAlreadyLocked) {
+            logger.info(`Proceeding as lock is already acquired for the resource: ${req.params.instance_id}`);
+          } else {
+            return next(err);
+          }
+        });
+    };
+  }
+
+  _handleWithUnlock(func, operationType, lastOperationCall, req, res, next) {
+    const fn = _.isString(func) ? this[func] : func;
+    return Promise
+      .try(() => fn.call(this, req, res))
+      .then(() => {
+        _.set(req, 'params_copy', req.params);
+        // if sf20 is enabled Check res status and unlock based on the request and status        
+        if (req.manager.name === CONST.INSTANCE_TYPE.DIRECTOR && config.enable_service_fabrik_v2) {
+          if (
+            operationType === CONST.OPERATION_TYPE.CREATE && res.statusCode === CONST.HTTP_STATUS_CODE.CONFLICT || // PutInstance => unlock in case of 409
+            operationType === CONST.OPERATION_TYPE.DELETE && res.statusCode === CONST.HTTP_STATUS_CODE.GONE || // DeleteInstance => unlock in case of 410
+            // lastoperation => unlock in case of status == 200 && state == 'succeeded' || state == 'failed'
```

Comment:
instead of these specific checks on method names and specific http resp codes, if the method handleWithResourceLock is invoked then response code should be 200 or 202 right? so cant these checks be made just to match these generic checks? if you have locked resource piror to invoking the method,  unlock it if response code is not 202.
I do see some code releated to unlock during last operation, but dont see lastOperation route mapped with this handler.

Reply:
I'll have to check if this will be always valid and test. Will do it as part of next PR.